rclrs/src/context/builder.rs

```diff
@@ -0,0 +1,148 @@
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::sync::{Arc, Mutex};
+
+use crate::rcl_bindings::*;
+use crate::{Context, RclrsError, ToResult};
+
+/// A builder for creating a [`Context`][1].
+///
+/// The builder pattern allows selectively setting some fields, and leaving all others at their default values.
+/// This struct instance can be created via [`Context::builder()`][2].
+///
+/// # Example
+/// ```
+/// # use rclrs::{Context, ContextBuilder, RclrsError};
+/// // Building a context in a single expression
+/// let args = ["ROS 1 ROS 2"].map(String::from);
+/// let context = ContextBuilder::new(args.clone()).domain_id(1).build()?;
+/// assert_eq!(context.domain_id(), 1);
+/// // Building a context via Context::builder()
+/// let context = Context::builder(args.clone()).domain_id(2).build()?;
+/// assert_eq!(context.domain_id(), 2);
+/// // Building a context step-by-step
+/// let mut builder = Context::builder(args.clone());
+/// builder = builder.domain_id(3);
+/// let context = builder.build()?;
+/// assert_eq!(context.domain_id(), 3);
+/// # Ok::<(), RclrsError>(())
+/// ```
+///
+/// [1]: crate::Context
+/// [2]: crate::Context::builder
+pub struct ContextBuilder {
+    arguments: Vec<String>,
+    domain_id: usize,
+}
+
+impl ContextBuilder {
+    /// Creates a builder for a context with arguments.
+    ///
+    /// Usually, this would be called with `std::env::args()`, analogously to `rclcpp::init()`.
+    /// See also the official "Passing ROS arguments to nodes via the command-line" tutorial.
+    ///
+    /// Creating a context can fail in case the args contain invalid ROS arguments.
+    ///
+    /// # Example
+    /// ```
+    /// # use rclrs::ContextBuilder;
+    /// let invalid_remapping = ["--ros-args", "-r", ":=:*/]"].map(String::from);
+    /// assert!(ContextBuilder::new(invalid_remapping).build().is_err());
+    /// let valid_remapping = ["--ros-args", "--remap", "__node:=my_node"].map(String::from);
+    /// assert!(ContextBuilder::new(valid_remapping).build().is_ok());
+    /// ```
+    pub fn new(args: impl IntoIterator<Item = String>) -> ContextBuilder {
+        let mut domain_id = 0;
+        // SAFETY: Getting the default domain ID, based on the environment
+        let ret = unsafe { rcl_get_default_domain_id(&mut domain_id) };
+        debug_assert_eq!(ret, 0);
+
+        ContextBuilder {
+            arguments: args.into_iter().collect(),
+            domain_id,
+        }
+    }
+
+    /// Sets the context domain id.
+    ///
+    /// The domain ID controls which nodes can send messages to each other, see the [ROS 2 concept article][1].
+    ///
+    /// [1]: https://docs.ros.org/en/rolling/Concepts/About-Domain-ID.html
+    pub fn domain_id(mut self, domain_id: usize) -> Self {
+        #[cfg(ros_distro = "foxy")]
+        std::env::set_var("ROS_DOMAIN_ID", domain_id.to_string());
```

Comment:
In Foxy, could you just disable the entire `domain_id` functionality in the context builder, and add it to the node builder instead?

Reply:
New commits haven't passed `doc-test` due to `domain_id` func doesn't exist in Foxy. I have searched [doc_cfg](https://doc.rust-lang.org/beta/unstable-book/language-features/doc-cfg.html) but it is not merged yet https://github.com/rust-lang/rust/issues/43781.
Is there any solutions to test each ros distro?

Reply:
I have modified `doc-test` :)